lib/senders/oauth_client_info.js

```diff
@@ -0,0 +1,91 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const request = require('request')
+const Keyv = require('keyv')
+const Joi = require('joi')
+
+const P = require('../promise')
+const config = require('../../config').getProperties()
+const log = require('./log')('oauth_client_info')
+
+const OAUTH_SERVER = config.oauth.url
+const OAUTH_CLIENT_INFO_CACHE_TTL = config.oauth.clientInfoCacheTTL
+const OAUTH_CLIENT_INFO_CACHE_NAMESPACE = 'oauthClientInfo'
+const FIREFOX_CLIENT = {
+  name: 'Firefox'
+}
+
+const clientCache = new Keyv({
+  ttl: OAUTH_CLIENT_INFO_CACHE_TTL,
+  namespace: OAUTH_CLIENT_INFO_CACHE_NAMESPACE
+})
+
+/**
+ * Fetches OAuth client info from the OAuth server.
+ * Stores the data into server memory.
+ * @param clientId
+ * @returns {Promise<any>}
+ */
+function fetch(clientId) {
+  log.trace({ op: 'fetch.start' })
+
+  const options = {
+    url: `${OAUTH_SERVER}/v1/client/${clientId}`,
+    method: 'GET',
+    json: true
+  }
+
+  return new P((resolve) => {
+    if (! clientId || clientId === 'sync') {
+      log.trace({ op: 'fetch.sync' })
+      return resolve(FIREFOX_CLIENT)
+    }
+
+    clientCache.get(clientId)
+      .then((cachedRecord) => {
+        if (cachedRecord) {
+          // used the cachedRecord
+          log.trace({ op: 'fetch.usedCache' })
+          return resolve(cachedRecord)
+        }
+
+        // request info from the OAuth server
+        request(options, function (err, res, body) {
+          if (err || res.statusCode !== 200 || ! body.name) {
+            if (err) {
+              log.critical({ op: 'fetch.failed', err: err })
+            } else {
+              log.warn({ op: 'fetch.failedForClient', clientId: clientId, name: !! body.name })
+            }
+            // fallback to the Firefox client if request fails
+            return resolve(FIREFOX_CLIENT)
+          }
+
+          const validation = Joi.validate(body.name, Joi.string().max(256).required())
```

Comment:
I suggest applying our existing `DISPLAY_SAFE_UNICODE` validator here, which should reject things that might be troublesome in an email header.  It's probably worthwhile to use that in oauth-server as well, if we don't already do so.

Reply:
Also, +1 to doing validation at this point, it feels right to me.  Since we're fetching external data that's under a small amount of influence from a potential attacker, we should treat it as hostile and enforce expectations on the boundary here in the same way we do for inbound requests.